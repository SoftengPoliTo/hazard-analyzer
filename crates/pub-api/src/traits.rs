use std::path::Path;

use rustdoc_types::{Crate, ItemEnum};
use serde::Serialize;

// Trait.
#[derive(Serialize)]
pub(crate) struct Trait<'a> {
    // Name.
    name: &'a str,
    // File.
    file: &'a Path,
    // Functions.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    functions: Vec<&'a str>,
}

pub(crate) fn get_trait<'a>(
    name: &'a str,
    file: &'a Path,
    trait_: &'a rustdoc_types::Trait,
    doc_crate: &'a Crate,
) -> Trait<'a> {
    // Get trait functions by getting the name of all
    // items of type Function contained in trait.items.
    let functions: Vec<_> = trait_
        .items
        .iter()
        .filter_map(|item| {
            doc_crate
                .index
                .get(item)
                .and_then(|item| match item.inner {
                    ItemEnum::Function(_) => item.name.as_ref(),
                    _ => None,
                })
                .map(|name| name.as_str())
        })
        .collect();

    Trait {
        name,
        file,
        functions,
    }
}
