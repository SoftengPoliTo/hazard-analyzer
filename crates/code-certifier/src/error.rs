//! This module handles errors.

use std::borrow::Cow;

/// All possible error kinds.
#[derive(Debug, Copy, Clone)]
pub enum ErrorKind {
    /// I/O error.
    Io,
    /// Git error.
    Git,
    /// Toml error.
    Toml,
    /// Rustdoc error.
    Rustdoc,
    /// Concurrent error.
    Concurrent,
    /// Json error.
    Json,
}

impl ErrorKind {
    pub(crate) const fn description(self) -> &'static str {
        match self {
            ErrorKind::Io => "I/O error",
            ErrorKind::Git => "Git error",
            ErrorKind::Toml => "TOML error",
            ErrorKind::Rustdoc => "Rustdoc error",
            ErrorKind::Concurrent => "Concurrent error",
            ErrorKind::Json => "JSON error",
        }
    }
}

impl std::fmt::Display for ErrorKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.description().fmt(f)
    }
}

/// Library error.
#[derive(Debug)]
pub struct Error {
    kind: ErrorKind,
    info: Cow<'static, str>,
}

impl Error {
    /// Creates a new [`Error`] instance.
    pub fn new(kind: ErrorKind, info: impl Into<Cow<'static, str>>) -> Self {
        Self {
            kind,
            info: info.into(),
        }
    }

    pub(crate) fn error(&self) -> String {
        format!("{}: {}", self.kind, self.info)
    }
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.error().fmt(f)
    }
}

impl From<std::io::Error> for Error {
    fn from(e: std::io::Error) -> Self {
        Self::new(ErrorKind::Io, e.to_string())
    }
}

impl<T> From<crossbeam::channel::SendError<T>> for Error {
    fn from(_: crossbeam::channel::SendError<T>) -> Self {
        Self::new(ErrorKind::Concurrent, "Sender error")
    }
}

impl From<serde_json::Error> for Error {
    fn from(e: serde_json::Error) -> Self {
        Self::new(ErrorKind::Json, e.to_string())
    }
}

impl From<git2::Error> for Error {
    fn from(e: git2::Error) -> Self {
        Self::new(ErrorKind::Git, e.to_string())
    }
}

impl From<toml::de::Error> for Error {
    fn from(e: toml::de::Error) -> Self {
        Self::new(ErrorKind::Toml, e.to_string())
    }
}

impl From<rustdoc_json::BuildError> for Error {
    fn from(e: rustdoc_json::BuildError) -> Self {
        Self::new(ErrorKind::Rustdoc, e.to_string())
    }
}

/// A specialized `Result` type.
pub type Result<T> = std::result::Result<T, Error>;
