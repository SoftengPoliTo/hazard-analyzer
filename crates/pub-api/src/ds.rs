use std::{
    collections::{HashMap, HashSet},
    path::Path,
};

use rustdoc_types::{Crate, Id, ItemEnum};
use serde::Serialize;

// Data structure impl.
//
// Can be a data structure implementation or
// a trait implementation.
enum Impl<'a> {
    // Data structure implementation with associated functions.
    // The key corresponds to the id of the function inside json rustdoc.
    Functions(HashMap<&'a Id, &'a str>),
    // Trait implementation.
    Trait(TraitImpl<'a>),
}

// Trait implementaion.
#[derive(Serialize)]
pub(crate) struct TraitImpl<'a> {
    // Trait name.
    name: &'a str,
    // Trait functions.
    // Contains the provided (with deafult implementation) functions
    // and the ones that have to be implemented (required).
    #[serde(skip_serializing_if = "Vec::is_empty")]
    functions: Vec<&'a str>,
}

// Data structure.
pub(crate) struct DataStructure<'a> {
    // Name.
    pub(crate) name: &'a str,
    // File.
    pub(crate) file: &'a Path,
    // Implemented functions.
    pub(crate) functions: HashMap<&'a Id, &'a str>,
    // Implemented traits.
    pub(crate) traits: Vec<TraitImpl<'a>>,
}

// Data structure representation inside output manifest.
#[derive(Serialize)]
pub(crate) struct ApiDataStructure<'a> {
    // Name.
    pub(crate) name: &'a str,
    // File.
    pub(crate) file: &'a Path,
    // Implemented functions.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub(crate) functions: Vec<&'a str>,
    // Implemented traits.
    pub(crate) traits: Vec<TraitImpl<'a>>,
}

impl<'a> From<DataStructure<'a>> for ApiDataStructure<'a> {
    fn from(data_structure: DataStructure<'a>) -> ApiDataStructure<'a> {
        Self {
            name: data_structure.name,
            file: data_structure.file,
            functions: data_structure.functions.values().copied().collect(),
            traits: data_structure.traits,
        }
    }
}

pub(crate) fn api_data_structures(data_structures: Vec<DataStructure>) -> Vec<ApiDataStructure> {
    data_structures.into_iter().map(|ds| ds.into()).collect()
}

// Retrieves the functions associated to an impl instance.
// They can be the functions related to a data structure implementation
// or the ones related to a trait impl by a data structure.
fn get_functions<'a, T, M, C>(doc_crate: &'a Crate, impl_: &'a rustdoc_types::Impl, map: M) -> C
where
    M: Fn(&'a Id, &'a str) -> T,
    C: FromIterator<T>,
{
    impl_
        .items
        .iter()
        .filter_map(|id| {
            let item = doc_crate.index.get(id)?;
            match item.inner {
                ItemEnum::Function(_) => {
                    let func_name = item.name.as_ref()?;
                    Some(map(&item.id, func_name))
                }
                _ => None,
            }
        })
        .collect()
}

// Retrieves functions defined in a data structure impl block.
fn ds_functions<'a>(
    doc_crate: &'a Crate,
    impl_: &'a rustdoc_types::Impl,
) -> HashMap<&'a Id, &'a str> {
    get_functions(doc_crate, impl_, |id, func_name| (id, func_name))
}

fn trait_functions<'a>(doc_crate: &'a Crate, impl_: &'a rustdoc_types::Impl) -> Vec<&'a str> {
    // Required functions and reimplementations of provided functions.
    let mut functions: HashSet<&str> = get_functions(doc_crate, impl_, |_, func_name| func_name);

    // Extend the functions list with the functions provided by the Trait that have not been reimplemented.
    functions.extend(impl_.provided_trait_methods.iter().map(|s| s.as_str()));

    functions.into_iter().collect()
}

fn get_impl<'a>(doc_crate: &'a Crate, impl_: &'a rustdoc_types::Impl) -> Impl<'a> {
    // Check if impl corresponds to a data structure implementation or to a trait one,
    // and create the corresponding Impl object.
    if let Some(trait_) = &impl_.trait_ {
        let functions = trait_functions(doc_crate, impl_);
        Impl::Trait(TraitImpl {
            name: &trait_.name,
            functions,
        })
    } else {
        let functions = ds_functions(doc_crate, impl_);
        Impl::Functions(functions)
    }
}

// Get implemented functions from array of data structure implementations.
fn implemented_functions(functions: Vec<Impl>) -> HashMap<&Id, &str> {
    functions
        .into_iter()
        .filter_map(|impl_| match impl_ {
            Impl::Functions(functions) => Some(functions),
            _ => None,
        })
        .flatten()
        .collect()
}

// Get implemented traits from array of data structure implementations.
fn implemented_traits(traits: Vec<Impl>) -> Vec<TraitImpl> {
    traits
        .into_iter()
        .filter_map(|impl_| match impl_ {
            Impl::Trait(trait_) => Some(trait_),
            _ => None,
        })
        .collect()
}

// Returns map of data structure implemented functions and array
// of implemented traits.
fn get_impls<'a>(
    doc_crate: &'a Crate,
    impls: &'a [Id],
) -> (HashMap<&'a Id, &'a str>, Vec<TraitImpl<'a>>) {
    // Retrieve impl.
    let retrieve_impl = |impl_: &Id| -> Option<Impl> {
        doc_crate
            .index
            .get(impl_)
            .and_then(|item| match &item.inner {
                ItemEnum::Impl(impl_item) => Some(get_impl(doc_crate, impl_item)),
                _ => None,
            })
    };

    // Distinguish between data structure implemented functions and implemented traits.
    let (functions, traits) = impls
        .iter()
        .filter_map(retrieve_impl)
        .partition::<Vec<_>, _>(|impl_| matches!(impl_, Impl::Functions(_)));

    // Get implemented functions.
    let functions = implemented_functions(functions);

    // Get implemented traits.
    let traits = implemented_traits(traits);

    (functions, traits)
}

pub(crate) fn get_data_structure<'a>(
    name: &'a str,
    file: &'a Path,
    impls: &'a [Id],
    doc_crate: &'a Crate,
) -> DataStructure<'a> {
    let (functions, traits) = get_impls(doc_crate, impls);

    DataStructure {
        name,
        file,
        functions,
        traits,
    }
}
