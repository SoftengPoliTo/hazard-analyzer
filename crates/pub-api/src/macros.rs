use std::path::Path;

use serde::Serialize;

// Macro.
#[derive(Serialize)]
pub(crate) struct Macro<'a> {
    // Name.
    pub(crate) name: &'a str,
    // File.
    pub(crate) file: &'a Path,
}
