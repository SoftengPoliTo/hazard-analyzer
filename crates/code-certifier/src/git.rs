//! This module handles the cloning of `ascot-firmware`.

use std::{env::temp_dir, path::PathBuf};

use git2::Repository;
use toml::Value;

use crate::error::{Error, ErrorKind, Result};

const CARGO_TOML: &str = include_str!("../Cargo.toml");

/// Clones ascot-firmware (https://github.com/SoftengPoliTo/ascot-firmware)
/// inside the temporary directory and returns the path to it.
pub fn ascot_firmware() -> Result<PathBuf> {
    // Parse Cargo.toml file.
    let cargo_toml: Value = CARGO_TOML.parse()?;

    // Get ascot-firmware repository URL.
    let repo_url = cargo_toml
        .get("package")
        .and_then(|pkg| pkg.get("metadata"))
        .and_then(|metadata| metadata.get("ascot-firmware"))
        .and_then(|ascot_firmware| ascot_firmware.get("url"))
        .and_then(|url| url.as_str())
        .ok_or(Error::new(
            ErrorKind::Git,
            "Invalid ascot-firmware repository URL Cargo.toml",
        ))?;

    // Clone ascot-firmware in the temporary directory.
    let repo_path = temp_dir().join("ascot-firmware");
    if repo_path.exists() {
        std::fs::remove_dir_all(&repo_path)?;
    }
    Repository::clone(repo_url, &repo_path)?;

    Ok(repo_path)
}

/// Returns `ascot-firmware/ascot-axum/src/devices` path.
#[inline(always)]
pub fn ascot_devices(ascot_firmware: PathBuf) -> PathBuf {
    ascot_firmware
        .join("ascot-axum")
        .join("src")
        .join("devices")
}
