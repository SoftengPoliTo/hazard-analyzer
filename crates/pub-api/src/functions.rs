use std::{collections::HashSet, path::Path};

use rustdoc_types::{Crate, Id, Item, ItemEnum, Visibility};
use serde::Serialize;

use super::ds::DataStructure;

// Function.
#[derive(Serialize)]
pub(crate) struct Function<'a> {
    // Name.
    name: &'a str,
    // File.
    file: &'a Path,
}

fn get_function<'a>(id: &Id, item: &'a Item, ds_functions: &HashSet<&Id>) -> Option<Function<'a>> {
    // Filter out the item if it is not from local crate OR if it is not public OR
    // if it is a function from a data structure impl.
    if item.crate_id != 0 || item.visibility != Visibility::Public || ds_functions.contains(&id) {
        return None;
    }

    // Check if item is actually a function and build the corresponding Function object.
    match item.inner {
        ItemEnum::Function(_) => {
            let function_name = item.name.as_ref()?;
            let file = &item.span.as_ref()?.filename;
            Some(Function {
                name: function_name.as_str(),
                file,
            })
        }
        _ => None,
    }
}

// Retrieves public functions by taking all crate public functions
// and filtering out those that are not linked to a Trait or to the impl
// of a data structure.
//
// The JSON output of rustdoc does not in fact provide a direct way
// to distinguish between plain functions and functions contained in impl blocks.
pub(crate) fn get_functions<'a>(
    doc_crate: &'a Crate,
    structs: &[DataStructure<'a>],
    enums: &[DataStructure<'a>],
) -> Vec<Function<'a>> {
    // Functions to filter out.
    let ds_functions: HashSet<_> = structs
        .iter()
        .chain(enums.iter())
        .flat_map(|ds| ds.functions.keys())
        .copied()
        .collect();

    // Get functions.
    doc_crate
        .index
        .iter()
        .filter_map(|(id, item)| get_function(id, item, &ds_functions))
        .collect()
}
