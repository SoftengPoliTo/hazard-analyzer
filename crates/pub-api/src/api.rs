use std::path::{Path, PathBuf};

use crate::ds::{api_data_structures, get_data_structure, ApiDataStructure};
use crate::functions::{get_functions, Function};
use crate::macros::Macro;
use crate::traits::{get_trait, Trait};
use rustdoc_types::{Crate, ItemEnum, Visibility};
use serde::Serialize;

use code_certifier::error::{Error, ErrorKind, Result};

use code_certifier::git::ascot_firmware;

const LIB_TOML_PATH: &str = "Cargo.toml";
const AXUM_TOML_PATH: &str = "ascot-axum/Cargo.toml";

// APIs.
#[derive(Serialize)]
struct Api<'a> {
    // Structs.
    structs: Vec<ApiDataStructure<'a>>,
    // Enums.
    enums: Vec<ApiDataStructure<'a>>,
    // Traits.
    traits: Vec<Trait<'a>>,
    // Functions.
    functions: Vec<Function<'a>>,
    // Macros.
    macros: Vec<Macro<'a>>,
}

// Public APIs.
#[derive(Serialize)]
pub(crate) struct PublicApis<'a> {
    // ascot-library public APIs.
    ascot_library: Api<'a>,
    // ascot-axum public APIs.
    ascot_axum: Api<'a>,
}

fn ascot_paths(
    library_path: Option<PathBuf>,
    axum_path: Option<PathBuf>,
) -> Result<(PathBuf, PathBuf)> {
    match (library_path, axum_path) {
        (None, None) => {
            let ascot_firmware = ascot_firmware()?;
            Ok((
                ascot_firmware.join(LIB_TOML_PATH),
                ascot_firmware.join(AXUM_TOML_PATH),
            ))
        }
        (None, Some(axum_path)) => {
            let ascot_firmware = ascot_firmware()?;
            Ok((ascot_firmware.join(LIB_TOML_PATH), axum_path))
        }
        (Some(library_path), None) => {
            let ascot_firmware = ascot_firmware()?;
            Ok((library_path, ascot_firmware.join(AXUM_TOML_PATH)))
        }
        (Some(library_path), Some(axum_path)) => Ok((library_path, axum_path)),
    }
}

fn rustdoc_crate<M: AsRef<Path>>(manifest_path: M) -> Result<Crate> {
    // Create json rustdoc for the project corresponding to the given manifest_path.
    let doc_path = rustdoc_json::Builder::default()
        .toolchain("nightly")
        .manifest_path(manifest_path)
        .build()?;

    let doc = std::fs::read_to_string(doc_path)?;

    // Extract the documentation Crate.
    let doc_crate: Crate = serde_json::from_str(&doc)?;

    Ok(doc_crate)
}

pub(crate) fn doc_crates(
    library_path: Option<PathBuf>,
    axum_path: Option<PathBuf>,
) -> Result<(Crate, Crate)> {
    let (library_path, axum_path) = ascot_paths(library_path, axum_path)?;

    Ok((rustdoc_crate(library_path)?, rustdoc_crate(axum_path)?))
}

fn get_api(doc_crate: &Crate) -> Api {
    let mut structs = vec![];
    let mut enums = vec![];
    let mut traits = vec![];
    let mut macros = vec![];

    // Explore one by one the crate public items and create the corresponding objects.
    doc_crate
        .index
        .iter()
        .filter(|(_, item)| item.crate_id == 0 && item.visibility == Visibility::Public)
        .for_each(|(_, item)| {
            if let (Some(name), Some(span)) = (&item.name, &item.span) {
                match &item.inner {
                    ItemEnum::Struct(struct_item) => {
                        let struct_ =
                            get_data_structure(name, &span.filename, &struct_item.impls, doc_crate);
                        structs.push(struct_);
                    }
                    ItemEnum::Enum(enum_item) => {
                        let enum_ =
                            get_data_structure(name, &span.filename, &enum_item.impls, doc_crate);
                        enums.push(enum_);
                    }
                    ItemEnum::Trait(trait_item) => {
                        let trait_ = get_trait(name, &span.filename, trait_item, doc_crate);
                        traits.push(trait_);
                    }
                    ItemEnum::Macro(_) => macros.push(Macro {
                        name,
                        file: &span.filename,
                    }),
                    _ => (),
                }
            }
        });

    // Retrieve the public functions of the project.
    let functions = get_functions(doc_crate, &structs, &enums);

    // Convert structs and enums to the manifest representation.
    let structs = api_data_structures(structs);
    let enums = api_data_structures(enums);

    Api {
        structs,
        enums,
        traits,
        functions,
        macros,
    }
}

pub(crate) fn apis<'a>(library_crate: &'a Crate, axum_crate: &'a Crate) -> PublicApis<'a> {
    let library_apis = get_api(library_crate);
    let axum_apis = get_api(axum_crate);

    PublicApis {
        ascot_library: library_apis,
        ascot_axum: axum_apis,
    }
}

pub(crate) fn check_ascot_path(ascot_path: &Path) -> Result<()> {
    // Check if ascot_path is a path to a .toml manifest.
    if ascot_path
        .extension()
        .map_or(false, |ext| ext.to_ascii_lowercase() != "toml")
    {
        return Err(Error::new(
            ErrorKind::Io,
            "Ascot path must be a path to the .toml manifest.",
        ));
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use insta::sorted_redaction;

    use super::{get_api, rustdoc_crate};

    const SNAPSHOTS_PATH: &str = "../tests/snapshots/";

    fn test_apis(snapshot_name: &str, manifest_path: &Path) {
        let doc_crate = rustdoc_crate(manifest_path).unwrap();
        let apis = get_api(&doc_crate);

        insta::with_settings!({
            snapshot_path => Path::new(SNAPSHOTS_PATH),
            prepend_module_to_snapshot => false,
            sort_maps => true
        }, {
            insta::assert_yaml_snapshot!(snapshot_name, apis,
                {
                    ".structs" => sorted_redaction(),
                    ".structs.*.functions" => sorted_redaction(),
                    ".structs.*.traits" => sorted_redaction(),
                    ".structs.*.traits.*.functions" => sorted_redaction(),
                    ".enums" => sorted_redaction(),
                    ".enums.*.functions" => sorted_redaction(),
                    ".enums.*.traits" => sorted_redaction(),
                    ".enums.*.traits.*.functions" => sorted_redaction(),
                    ".traits" => sorted_redaction(),
                    ".traits.*.functions" => sorted_redaction(),
                    ".functions" => sorted_redaction(),
                    ".macros" => sorted_redaction()
                })
        });
    }

    #[test]
    fn structs() {
        test_apis("structs", Path::new("./tests/projects/structs/Cargo.toml"));
    }

    #[test]
    fn enums() {
        test_apis("enums", Path::new("./tests/projects/enums/Cargo.toml"));
    }

    #[test]
    fn traits() {
        test_apis("traits", Path::new("./tests/projects/traits/Cargo.toml"));
    }

    #[test]
    fn functions() {
        test_apis(
            "functions",
            Path::new("./tests/projects/functions/Cargo.toml"),
        );
    }

    #[test]
    fn macros() {
        test_apis("macros", Path::new("./tests/projects/macros/Cargo.toml"));
    }

    #[test]
    fn impl_traits() {
        test_apis(
            "impl_traits",
            Path::new("./tests/projects/impl_traits/Cargo.toml"),
        );
    }

    #[test]
    fn derive_traits() {
        test_apis(
            "derive_traits",
            Path::new("./tests/projects/derive_traits/Cargo.toml"),
        );
    }

    #[test]
    fn aliases() {
        test_apis("aliases", Path::new("./tests/projects/aliases/Cargo.toml"));
    }
}
