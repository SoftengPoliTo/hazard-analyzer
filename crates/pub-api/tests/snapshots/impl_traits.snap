---
source: crates/pub-api/src/api.rs
expression: apis
---
structs:
  - name: Struct
    file: crates/pub-api/tests/projects/impl_traits/src/lib.rs
    traits:
      - name: Any
        functions:
          - type_id
      - name: Borrow
        functions:
          - borrow
      - name: BorrowMut
        functions:
          - borrow_mut
      - name: Freeze
      - name: From
        functions:
          - from
      - name: Into
        functions:
          - into
      - name: RefUnwindSafe
      - name: Send
      - name: Sync
      - name: Trait1
        functions:
          - provided_function1
          - required_function1
      - name: Trait2
        functions:
          - provided_function2
          - required_function2
      - name: TryFrom
        functions:
          - try_from
      - name: TryInto
        functions:
          - try_into
      - name: Unpin
      - name: UnwindSafe
enums: []
traits:
  - name: Trait1
    file: crates/pub-api/tests/projects/impl_traits/src/lib.rs
    functions:
      - provided_function1
      - required_function1
  - name: Trait2
    file: crates/pub-api/tests/projects/impl_traits/src/lib.rs
    functions:
      - provided_function2
      - required_function2
functions: []
macros: []
